src/experimental/Db.h

```diff
@@ -157,18 +162,18 @@ class DB {
      * \brief Verify entry data is sane for the database.
      */
     static
-    bool verify(Entry entry)
+    Entry prepareDataForCommit(Entry entry)
+    /// [F] this function does not really verify data, it just removes bad entries,
+    /// maybe returning the 'cleaned up' entry would be better?
```

Comment:
Im guessing this is an internal function to cleanup the to-be-updated entry.
However where does the "commit" come in? Is it going to just be a wrapper around insert and update?
If so what will be checked for whether Entry is to be updated or inserted?

Reply:
That's correct. I would implement commit something like this (once prepareDataForCommit is finished):
```cpp
bool commit(Entry entry)
{
    if (exists(entry)) {
        return update(entry);
    } else {
        return insert(entry);
    }
}
```
exists would check the position member of the entry object. The default position for a new object is {"table",0}, whereas an existing entry has a rowId != 0. However, if not 0, we could (should?) also confirm the entry exists in the database. This adds some overhead but might be worth it. If we ommit this check, exists would only have to check position.second

```cpp
bool exists(Entry entry)
{
    if (entry.position.second == 0)
        return false;
    
    //Check database for row id
    QString statement = "SELECT COUNT(*) FROM " + entry.position.first +
                        " WHERE _rowid_=" + QString::number(entry.position.second);
    //this returns either 1 or 0 since row ids are unique
    QSqlQuery q(statement);
    q.next();
    int rowId = q.value(0).toInt();
    if (rowId) {
        DEB("Entry exists with row ID: " << rowId);
        return true;
    } else {
        DEB("Entry does not exist.");
        return false;
    }
}
```

Reply:
Fair enough this makes sense but api wise how will it work?
For example the NewPilotDialog. what will it call?
If the user creates new entry its easy bcause we create stuff from the UserInput
but if he updates an entry? Arent we supposed to just call the update after cleaning the entry?

Reply:
Well, the 'clean-up' function would be called before updating or inserting, since it has to be called for both new entries as well as edited entries maybe we can call it in the constructor for entry?

The flow could look something like:

NewPilot Dialog is ready and calls
```cpp
UserInput rawData =  collectInput(QList<QLineEdit*>);

Entry toDatabase = Entry(rawData);
// constructor of entry reads metadata and calls clean up function

if (editRole == Db::editExisting)
    toDatabase.setPosition(oldPosition)
// set position if editing a new entry, otherwise default position will be 0

DB::commit(toDatabase);
// depending on position data either calls insert or update

// on success, accept dialog, otherwise prompt user.
```
What do you think?

Reply:
Im not convinced this is a good course of action. It feels like db logic is spilling out again. If the user is editing an entry then simply by clicking it from the gui we have the position in the database so we can directly call `update(updated_entry)` (and being explicit in our intent). I think im missing something. 

We need to clarify this distiction of updating vs inserting not only here in db but also gui level. If by clicking the existing pilots the user is selecting Entries then the dialog should edit that `Entry` and pass it to the database via `update(Entry)`.
We could keep an `Entry` encapsulated in the dialog. If we call the overloaded constructor then we keep the entry which gets updated after the user is done editing and pass that to the `db::update(Entry)`. We can keep the edit role if you think it makes the potential outcome of the dialog explicit (which i would agree with)
```c++
class NewPilotDialog {
    ...
    Entry potential_entry;
    ...
};

void NewPilotDialog::submitForm()
{
    switch (role) {
    case Db::createNew:
        if(!DB::insert(uin)) emit QDialog::rejected();
        break;
    case Db::editExisting:
        if(!DB::update(potential_entry)) emit QDialog::rejected();
        break;
    }
}
```
It needs a bit of refactoring but i think it still keeps what we are trying to achieve with cleaning up the responsibilities

Reply:
In this case, what would be the distinction between UserInput and Entry? Wasn't the idea to have the entry represent a 'sane' database entry and the userInput as a 'raw' data that needs some processing before it is being fed to the database?

Keeping the old entry around in the dialog would not fix the need for cleanup either. This step is not unique to a new entry. Let me  give you an example:
Data is collected from the dialog like this:
```cpp
void NewPilotDialog::submitForm()
{
...
    auto line_edits = parent()->findChildren<QLineEdit *>();

    for (const auto &le : line_edits) {
        QString key = le->objectName();
        key.chop(8);//remove "LineEdit"
        QString value = le->text();
        if (!key.isEmpty()) {
            newData.insert(key, value);
            DEB("Data entry: " << key << value);
        }
    }
...
}
```
This gives us the following data:
```cpp
void NewPilotDialog::submitForm() 	 Creating Database Object...
void NewPilotDialog::submitForm() 	 Data entry:  "search" ""
void NewPilotDialog::submitForm() 	 Data entry:  "picfirstname" "Adam"
void NewPilotDialog::submitForm() 	 Data entry:  "piclastname" "Knox"
void NewPilotDialog::submitForm() 	 Data entry:  "company" "Lufthansa"
void NewPilotDialog::submitForm() 	 Data entry:  "alias" ""
void NewPilotDialog::submitForm() 	 Data entry:  "employeeid" ""
void NewPilotDialog::submitForm() 	 Data entry:  "phone" ""
void NewPilotDialog::submitForm() 	 Data entry:  "email" ""
````
In this example, cleanUp would remove the entry **{"search" ""}**, as it originates from searchLineEdit and is not part of the database entry.

Now, if we would use this data as is (without cleaning it up) on an 'old' entry with entry.setData(newData), the entry would be polluted and the update query would fail.

Reply:
I just looked at the pilot dialogue. You really dont need to sanatize the input. You *have* to do it now because you are looking for the line edits at the wrong place. You only need to change on line and there will be no "bad inputs"
```c++
// change this to
auto line_edits = parent()->findChildren<QLineEdit *>();
// this
auto line_edits = this->findChildren<QLineEdit*>();
```
![pilotdialog](https://user-images.githubusercontent.com/69913906/101477455-861a4400-394f-11eb-9f2a-b0ab07273232.png)

Once again the simplest solution is the best :D
Doesnt this clean up the db flow further? because now we can tweak the data in the encapsulated `Entry` and pass it in the `db::update(updated_entry)` and no sanatizing is necessary.
The dialog is constructed with the clicked `Entry` the user tweaks stuff we update the data in the `Entry` and pass it to `update`.

Reply:
Works for me, problem solved! I'll remove the cleanUp function entirely and merge then!